File: python/bowling/bowling.py
from functools import reduce

class BowlingGame(object):
    def __init__(self):
        self.frames = []

    def roll(self, pins):
        if pins > 10 or pins < 0:
            raise ValueError('invalid pins')
        if len(self.frames) > 10:
            if len(self.frames[9]) == 2 and len(self.frames[10]) >= 1:
                raise IndexError('Spare should not have two bonus')
            elif len(self.frames[9]) == 1 and len(self.frames[10]) >= 2:
                raise IndexError('Strike should not have more than two bonus')

        if len(self.frames) == 0 or len(self.frames[-1]) == 2 or sum(self.frames[-1]) == 10:
            # last frame was finished
            self.frames.append([pins])
        else:
            self.frames[-1].append(pins)
        
        if any(sum(frame) > 10 for frame in self.frames):
            raise ValueError('invalid score')
        if len(self.frames) > 10:
            if sum(self.frames[9]) != 10:
                raise IndexError('Frame 10 is not a Strike or Spare')

    def score(self):
        if len(self.frames) < 10:
            raise IndexError('frame length is less than 10')
        
        if sum(self.frames[9]) == 10:
            if len(self.frames) <= 10 or len([e for frame in self.frames[9:] for e in frame]) != 3:
                raise IndexError('Bonus frame is invalid')

        frame_scores = []
        for i in range(10):
            frame = self.frames[i]
            if sum(frame) == 10:
                if len(frame) == 1:
                    frame_scores.append(10 + sum([e for frame in self.frames[i+1:] for e in frame][:2]))
                else:
                    frame_scores.append(10 + self.frames[i+1][0])
            else:
                frame_scores.append(sum(frame))
        
        return sum(frame_scores)

File: ruby/bowling/bowling.rb
class Game

  attr_reader :score_list

  class BowlingError < StandardError; end

  def initialize
    @score_list = []
  end

  def roll(pins)
    # a pins should between 0 to 10
    raise BowlingError if pins > 10 || pins < 0

    # split the pins for each frame, special case for 10th frame
    if @score_list.empty?
      @score_list << [*pins]
    elsif @score_list[-1].length < 2 && @score_list[-1].reduce(:+) < 10 #spare or open frame
      @score_list[-1] << pins
    elsif @score_list[-1].length == 2 || @score_list[-1].reduce(:+) == 10 #previous frame is finished
        @score_list << [*pins]
    end

    raise BowlingError if @score_list.any? { |e| e.reduce(:+) > 10 } #ignore the 10th frame
    #frame count would be 11 or 12 if 10th frame is a strike/spare
    raise BowlingError if @score_list.length > 10 && @score_list[9].reduce(:+) != 10

  end

  def score
    frame_scores = []
    score_list_dup = @score_list.dup

    # unfinished game could not be scored
    raise BowlingError if @score_list.length < 10

    # exactly check the bonus frame
    if score_list_dup[9].reduce(:+) == 10
      raise BowlingError if score_list_dup[10].nil? || score_list_dup[9..-1].flatten.count != 3
    end

    while frame_scores.count < 10
      current_frame = score_list_dup.shift
      if current_frame.reduce(:+) == 10
        if current_frame.count == 1
          # strike, add next two throws
          frame_scores << 10 + score_list_dup.flatten[0..1].reduce(:+)
        else
          # spare, add next one throw
          frame_scores << 10 + score_list_dup.flatten[0]
        end
      else
        frame_scores << current_frame.reduce(:+)
      end
    end

    frame_scores.reduce(:+)

  end

end


module BookKeeping
  VERSION = 3
end

